File: cs50/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>


bool validated(long card);
int get_length(long card);
bool checksum(long card);
void checkType(long card);

int main(void)
{
    long credit_number;

    do
    {
        credit_number = get_long("Number: ");
    }
    while (credit_number < 0);

    if (validated(credit_number) == true)
    {
        checkType(credit_number);
    }
    else
    {
        printf("INVALID\n");
    }
}


bool validated(long card)
{
    // get length of card number
    int len = get_length(card);
    return (len == 13 || len == 15 || len == 16) && checksum(card);
}

int get_length(long card)
{
    int length;
    for (length = 0; card != 0; card /= 10, length++);
    return length;
}

bool checksum(long card)
{
    int total = 0;
    int digit;
    for (int i = 0; card != 0; i++, card /= 10)
    {
        if (i % 2 == 0)
        {
            // add sum to sum of remaining numbers
            total += card % 10;
        }
        else
        {
            // multiply every other number by 2
            digit = 2 * (card % 10);
            // add digits of every product + qdd to sum
            total += digit % 10 + digit / 10;
        }
    }
    // check final digit = 0
    return (total % 10) == 0;
}

void checkType(long card)
{
    if ((card >= 34e13 && card < 35e13) || (card >= 37e13 && card < 38e13))
    {
        printf("AMEX\n");
    }
    else if (card >= 51e14 && card < 56e14)
    {
        printf("MASTERCARD\n");
    }
    else if ((card >= 4e12 && card < 5e12) || (card >= 4e15 && card < 5e15))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: cs50/pset6/credit/credit.py
import re
from math import floor


def main():
    while True:
        try:
            credit_number = int(input("Number: "))
            if credit_number > 0:
                break
        except:
            print("")
    if validated(credit_number):
        checkType(credit_number)
    else:
        print("INVALID")


def validated(card):
    # get length of card number
    len = get_length(card)
    return (len == 13 or len == 15 or len == 16) and checksum(card)


def get_length(card):
    length = len(str(card))
    return length


def checksum(card):
    total = 0
    i = 0
    # for (int i = 0; card != 0; i++, card /= 10)
    while not card == 0:
        if i % 2 == 0:
            # add sum to sum of remaining numbers
            total += card % 10
        else:
            # multiply every other number by 2
            digit = 2 * (card % 10)
            # add digits of every product + qdd to sum
            total += digit % 10 + floor(digit / 10)
        i += 1
        card /= 10
        card = floor(card)
    # check final digit = 0
    return (total % 10) == 0


def checkType(card):
    s = str(card)
    if len(s) == 15:
        if re.search("^34*", s) or re.search("^37*", s):
            print("AMEX")
    elif len(s) == 16:
        if re.search("^51*", s) or re.search("^52*", s) or re.search("^53*", s) or re.search("^54*", s) or re.search("^55*", s):
            print("MASTERCARD")
        elif re.search("^4*", s):
            print("VISA")
    elif len(s) == 13:
        if re.search("^4*", s):
            print("VISA")
    else:
        print("INVALID")


main()